src/lib/openjp2/segmented_stream.c

```diff
@@ -0,0 +1,335 @@
+{
+	return OPJ_FALSE;
+}
+
+static OPJ_BOOL opj_seg_buf_add_segment(opj_seg_buf_t* seg_buf, OPJ_BYTE* buf, OPJ_SIZE_T len) {
+	opj_buffer_t* new_seg = NULL;
+	new_seg = (opj_buffer_t*)opj_malloc(sizeof(opj_buffer_t));
+	if (!new_seg)
+		return OPJ_FALSE;
+
+	memset(new_seg, 0, sizeof(opj_buffer_t));
+	new_seg->buf = buf;
+	new_seg->len = len;
+	opj_vec_append(&seg_buf->segments, new_seg);
+
+	seg_buf->cur_seg_id = seg_buf->segments.size - 1;
+	seg_buf->data_len += len;
+	return OPJ_TRUE;
+}
+
+/*--------------------------------------------------------------------------*/
+
+void opj_seg_buf_destroy(opj_seg_buf_t* seg_buf) {
+	OPJ_INT32 i;
+	if (!seg_buf)
+		return;
+	for (i = 0; i < seg_buf->segments.size; ++i) {
+		opj_buffer_t* seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, i);
+		if (seg)
+			opj_buffer_free(seg);
+	}
+	opj_vec_free(&seg_buf->segments);
+}
+
+void opj_seg_buf_rewind(opj_seg_buf_t* seg_buf) {
+	OPJ_INT32 i;
+	if (!seg_buf)
+		return;
+	for (i = 0; i < seg_buf->segments.size; ++i) {
+		opj_buffer_t* seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, i);
+		if (seg) {
+			seg->offset = 0;
+		}
+	}
+	seg_buf->cur_seg_id = 0;
+}
+
+
+OPJ_BOOL opj_seg_buf_push_seg(opj_seg_buf_t* seg_buf, OPJ_BYTE* buf, OPJ_SIZE_T len) {
+	opj_buffer_t* seg = NULL;
+	if (!seg_buf || !buf || !len)
+		return OPJ_FALSE;
+
+	if (!seg_buf->segments.data) {
+		opj_vec_init(&seg_buf->segments);
+		seg_buf->segments.owns_data = OPJ_FALSE;
+	}
+
+
+	if (!opj_seg_buf_add_segment(seg_buf, buf, len))
+		return OPJ_FALSE;
+
+	seg = (opj_buffer_t*)opj_vec_back(&seg_buf->segments);
+	seg->owns_data = OPJ_FALSE;
+	return OPJ_TRUE;
+}
+
+OPJ_BOOL opj_seg_buf_alloc_seg(opj_seg_buf_t* seg_buf, OPJ_SIZE_T len) {
+	opj_buffer_t* seg = NULL;
+	OPJ_BYTE* buf = NULL;
+	if (!seg_buf || !len)
+		return OPJ_FALSE;
+
+	if (!seg_buf->segments.data) {
+		opj_vec_init(&seg_buf->segments);
+		seg_buf->segments.owns_data = OPJ_FALSE;
+	}
+
+
+	buf = (OPJ_BYTE*)opj_malloc(len);
+	if (!buf)
+		return OPJ_FALSE;
+
+	if (!opj_seg_buf_add_segment(seg_buf, buf, len))
+		return OPJ_FALSE;
+
+	seg = (opj_buffer_t*)opj_vec_back(&seg_buf->segments);
+	seg->owns_data = OPJ_TRUE;
+
+	return OPJ_TRUE;
+}
+
+void opj_seg_buf_incr_cur_seg_offset(opj_seg_buf_t* seg_buf, OPJ_OFF_T offset) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return;
+	cur_seg = (opj_buffer_t*)(seg_buf->segments.data[seg_buf->cur_seg_id]);
+	opj_buffer_incr_offset(cur_seg, offset);
+	if (cur_seg->offset == cur_seg->len) {
+		opj_seg_buf_increment(seg_buf);
+	}
+
+}
+
+
+/**
+* Zero copy read of contiguous chunk from current segment.
+* Returns OPJ_FALSE if unable to get a contiguous chunk, OPJ_TRUE otherwise
+*/
+OPJ_BOOL opj_seg_buf_zero_copy_read(opj_seg_buf_t* seg_buf,
+										OPJ_BYTE** ptr,
+										OPJ_SIZE_T chunk_len) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return OPJ_FALSE;
+	cur_seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id);
```

Comment:
given the definition of `opj_vec_get`, `cur_seg` could be `NULL`. An assert would be nice here (or maybe in `opj_vec_get` for bounds checking).